### PR TITLE
create landmark assessment

### PR DESCRIPTION
**Pull request by ada-only-one-ada**

### CODE REVIEW

File: src/api/map/routes/map.ts [hunk 1]
```diff
@@ -14,7 +14,11 @@ import db from "../../database/db";
 import { 
   landmarkAssessmentAttributes 
 } from "../../database/models/map/attributesAndIncludes";
-import { landmarkAssessmentInclude } from "../../database/models/map/attributesAndIncludes";
+import { 
+  landmarkAssessmentInclude 
+} from "../../database/models/map/attributesAndIncludes";
```
Comment (weihanwang) [inline src/api/map/routes/map.ts:19]: merge into the above import

File: src/api/map/routes/map.ts [hunk 2]
```diff
@@ -39,6 +43,32 @@ const list = procedure
...
+  .use(authUser())
+  .input(z.object({
+    landmark: z.string(),
+    score: z.number(),
+    markdown: z.string().nullable(),
+  }))
+  .mutation(async ({ ctx, input }) => {
+    if (input.score === undefined) {
+      throw generalBadRequestError("选项不能为空");
+    }
```
Comment (weihanwang) [inline src/api/map/routes/map.ts:56]: you don't need this. zod checks type validity automatically for you.

File: src/api/map/routes/map.ts [hunk 3]
```diff
@@ -39,6 +43,32 @@ const list = procedure
...
+    
+    return await sequelize.transaction(async transaction => {
+      const { landmark, score, markdown } = input;
+      
+      const la = await db.LandmarkAssessment.create({
+        userId: ctx.user.id,
```
Comment (weihanwang) [inline src/api/map/routes/map.ts:62]: pass in the userId and do not use the current user id

File: src/api/map/routes/map.ts [hunk 4]
```diff
@@ -39,6 +43,32 @@ const list = procedure
...
+        landmark: landmark,
+        score: score,
+        markdown: markdown,
+      }, { transaction });
+
+      return la;
```
Comment (weihanwang) [inline src/api/map/routes/map.ts:68]: no need to return anything here.

File: src/components/LandmarkDrawer.tsx [hunk 5]
```diff
@@ -28,20 +28,22 @@ import { Landmark, LandmarkAssessment, LandmarkScore } from 'shared/Map';
...
 
 const desktopTextLimit = 60;
 const mobileTextLimit = 20;
 
 export default function LandmarkDrawer ({ onClose, landmark }: { 
   onClose: () => void; 
   landmark: Landmark
-}) {          
+}) {   
+  const [user] = useUserContext();
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:46]: push `user` down to LandmarkAssessment and LandmarkAssessmentHistory

File: src/components/LandmarkDrawer.tsx [hunk 6]
```diff
@@ -62,10 +64,31 @@ function LandmarkDefinition ({ definition }: { definition: string })  {
...
+function LandmarkAssessment ({ landmark, user }: {
   landmark: Landmark;
+  user: User;
 }) {
   const [score, setScore] = useState<LandmarkScore | undefined>();
+  const [markdown, setMarkdown] = useState<string>("");
+  const { refetch } = trpcNext.map.listLandmarkAssessment.useQuery({
+    userId: user.id,
+    landmark: landmark.名称,
+  });
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:76]: remove this block. you don't need it just yet.

File: src/components/LandmarkDrawer.tsx [hunk 7]
```diff
@@ -62,10 +64,31 @@ function LandmarkDefinition ({ definition }: { definition: string })  {
...
+  const createLandmarkAssessment = async() => {
+    try {
+      if (score !== undefined) {
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:79]: you should do `invariant(score !== undefined` instead

File: src/components/LandmarkDrawer.tsx [hunk 8]
```diff
@@ -62,10 +64,31 @@ function LandmarkDefinition ({ definition }: { definition: string })  {
...
+        await trpc.map.createLandmarkAssessment.mutate({
+          landmark: landmark.名称,
+          score: score + 1,
+          markdown: markdown,
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:83]: simply `markedown,`

File: src/components/LandmarkDrawer.tsx [hunk 9]
```diff
@@ -62,10 +64,31 @@ function LandmarkDefinition ({ definition }: { definition: string })  {
...
+        });
+        void refetch();
+      }
+    } finally {
+      setScore(undefined);
+      setMarkdown("");
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:89]: why do you need to reset them?

File: src/components/LandmarkDrawer.tsx [hunk 10]
```diff
@@ -118,7 +147,12 @@ function LandmarkAssessmentHistory({ landmark } : {
         <Th>详情</Th>
       </Tr>
       <Tbody>
-        {assessments?.map((assessment, index) => (
+        {assessments?.sort((a1, a2) => {
+          const date1 = new Date(a1.createdAt ?? "");
+          const date2 = new Date(a2.createdAt ?? "");
+          return date2.getTime() - date1.getTime();  
+        })
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:154]: extract the sorting function in src/components/ChatRoom.tsx:36 into a utility function in strings.ts and reuse this function.

File: src/components/LandmarkDrawer.tsx (hunk 9 above)
Comment (ada-only-one-ada) [reply src/components/LandmarkDrawer.tsx:89]: After submitting, the result values should be removed?
Here is a screenshot without reset them:
Screenshot 2024-10-08 at 10 18 39

File: src/shared/strings.ts [hunk 11]
```diff
@@ -50,6 +51,14 @@ export function diffInMinutes(from: Date | string, to: Date | string): number {
   return Math.floor((new Date(to).getTime() - new Date(from).getTime()) / 1000 / 60);
 }
 
+export function sortByDateDesc(contents: any[]) : void{
```
Comment (weihanwang) [inline src/shared/strings.ts:54]: change to `compareDate(d1: ..., d2: ...)`. do sorting in the call sites instead.

File: src/api/map/routes/map.ts [hunk 12]
```diff
@@ -39,6 +40,27 @@ const list = procedure
...
+      await db.LandmarkAssessment.create({
+        userId,
+        landmark,
+        score,
+        markdown,
+      }, { transaction });
+    });
+  });
+
 const listLandmarkAssessment = procedure
```
Comment (weihanwang) [inline src/api/map/routes/map.ts:64]: can you change it to `listLandmarkAssessments` either in this PR or a separate one?

File: src/components/ChatRoom.tsx [hunk 13]
```diff
@@ -32,11 +32,8 @@ export default function Room({ menteeId }: {
   return !room ? <Loader /> :
     <VStack spacing={paragraphSpacing * 1.5} align="start" maxWidth="800px">
       <MessageCreator roomId={room.id} />
-
-      {room.messages.sort((a, b) => moment(a.updatedAt)
-        .isAfter(moment(b.updatedAt)) ? -1 : 1)
-        .map(m => <Message key={m.id} message={m} />)
-      }
+      {room.messages.sort((a, b) => compareDate(a.updatedAt, b.updatedAt))
```
Comment (weihanwang) [inline src/components/ChatRoom.tsx:35]: actually I should have used `createdAt`. can you help the fix in this or a separate PR?

File: src/shared/strings.ts [hunk 14]
```diff
@@ -50,6 +51,10 @@ export function diffInMinutes(from: Date | string, to: Date | string): number {
   return Math.floor((new Date(to).getTime() - new Date(from).getTime()) / 1000 / 60);
 }
 
+export function compareDate(d1: string | undefined, d2: string | undefined) {
+  return moment(d1 || "").isAfter(moment(d2 || "")) ? -1 : 1;
```
Comment (weihanwang) [inline src/shared/strings.ts:55]: the logic should be reversed to comply with the convention of the `compare` functions

Comment (weihanwang) [inline src/shared/strings.ts:55]: Do not do `|| ""` just pass in undefined times as is.

Comment (weihanwang) [inline src/shared/strings.ts:55]: add a logic to return 0 on equal times.

File: src/components/LandmarkDrawer.tsx [hunk 15]
```diff
@@ -65,7 +67,18 @@ function LandmarkDefinition ({ definition }: { definition: string })  {
...
 }) {
+  const [user] = useUserContext();      
   const [score, setScore] = useState<LandmarkScore | undefined>();
+  const [markdown, setMarkdown] = useState<string>("");
+  const createLandmarkAssessment = async() => {
+    invariant(score !== undefined);
+    await trpc.map.createLandmarkAssessment.mutate({
+      userId: user.id,
+      landmark: landmark.名称,
+      score: score + 1,
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:78]: don't plus one here. do `value={String(index + 1)}` at line 88

File: src/components/LandmarkDrawer.tsx [hunk 16]
```diff
@@ -101,11 +120,12 @@ function getAssessmentDate(assessment: LandmarkAssessment) {
 function LandmarkAssessmentHistory({ landmark } : {
   landmark: Landmark;
 }) {  
-  const [user] = useUserContext();
+  const [user] = useUserContext();
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:123]: spaces. please make sure you always self review your PRs before sending me for review.

File: src/components/LandmarkDrawer.tsx [hunk 17]
```diff
@@ -118,7 +138,8 @@ function LandmarkAssessmentHistory({ landmark } : {
         <Th>详情</Th>
       </Tr>
       <Tbody>
-        {assessments?.map((assessment, index) => (
+        {assessments?.sort((a, b) => compareDate(a.createdAt, b.createdAt))
```
Comment (weihanwang) [inline src/components/LandmarkDrawer.tsx:141]: in a separate PR, show <Loader /> during loading.

File: src/components/ChatRoom.tsx [hunk 18]
```diff
@@ -32,11 +32,8 @@ export default function Room({ menteeId }: {
   return !room ? <Loader /> :
     <VStack spacing={paragraphSpacing * 1.5} align="start" maxWidth="800px">
       <MessageCreator roomId={room.id} />
-
-      {room.messages.sort((a, b) => moment(a.updatedAt)
-        .isAfter(moment(b.updatedAt)) ? -1 : 1)
-        .map(m => <Message key={m.id} message={m} />)
-      }
+      {room.messages.sort((a, b) => compareDate(a.createdAt, b.createdAt))
```
[the remark this answers is not in the document]
Comment (ada-only-one-ada) [reply src/components/ChatRoom.tsx:35]: Yes.
Screenshot 2024-10-10 at 18 32 45

File: src/components/LandmarkDrawer.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (ada-only-one-ada) [reply src/components/LandmarkDrawer.tsx:141]: Yes
Screenshot 2024-10-10 at 18 31 59

**Review by weihanwang — APPROVED** [review]

File: src/components/LandmarkDrawer.tsx [hunk 19]
```diff
@@ -65,24 +67,41 @@ function LandmarkDefinition ({ definition }: { definition: string })  {
...
 }) {
+  const [user] = useUserContext();      
   const [score, setScore] = useState<LandmarkScore | undefined>();
+  const [markdown, setMarkdown] = useState<string>("");
+  const createLandmarkAssessment = async() => {
+    invariant(score !== undefined);
+    await trpc.map.createLandmarkAssessment.mutate({
+      userId: user.id,
+      landmark: landmark.名称,
+      score: score,
```
[the remark this answers is not in the document]
Comment (ada-only-one-ada) [reply src/components/LandmarkDrawer.tsx:78]: Will do this in a seperate PR.

**Pull request merged**